### PR TITLE
Bluetooth: GATT: Client and testing enhancements

### PR DESCRIPTION
**Pull request by Vudentz**
These changes make GATT Client APIs simpler to use by making parameters to be allocated by the stack itself.

In addition to that it introduces support for loopback connection which is then used to validate the changes and creates more tests to cover more APIs, in the future this may allow qualification tests to be performed as well with a very minimal overhead/latency added to CI.

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -3083,24 +3110,25 @@ int bt_gatt_subscribe(struct bt_conn *conn,
...
+	ret = 0;
 
 	/*
 	 * Add subscription before write complete as some implementation were
 	 * reported to send notification before reply to CCC write.
 	 */
 	gatt_subscription_add(conn, params);
 
-	return 0;
+	/* Skip write if already subscribed */
```
Comment (carlescufi) [inline subsys/bluetooth/host/gatt.c]: Why did you move this below `gatt_subscription_add()`?

File: include/bluetooth/gatt.h [hunk 2]
```diff
@@ -811,7 +811,7 @@ struct bt_gatt_notify_params {
  *  @param conn Connection object.
  *  @param params Notification parameters.
  *
- *  @return 0 in case of success or negative value in case of error.
+ *  @return request id in case of success or negative value in case of error.
```
Comment (carlescufi) [inline include/bluetooth/gatt.h]: Can the application safely ignore the request id and keep working as it was prior to the introduction of IDs?

File: samples/bluetooth/central_hr/src/main.c [hunk 3]
```diff
@@ -82,7 +82,7 @@ static u8_t discover_func(struct bt_conn *conn,
 		subscribe_params.ccc_handle = attr->handle;
 
 		err = bt_gatt_subscribe(conn, &subscribe_params);
-		if (err && err != -EALREADY) {
+		if (err < 0 && err != -EALREADY) {
```
Comment (carlescufi) [inline samples/bluetooth/central_hr/src/main.c]: Some apps out there are going to have to adapt to this. We should somehow make it visible

File: subsys/bluetooth/host/conn.c [hunk 4]
```diff
@@ -1954,7 +1972,8 @@ struct bt_conn *bt_conn_create_le(const bt_addr_le_t *peer,
 {
 	struct bt_conn *conn;
 
-	if (!atomic_test_bit(bt_dev.flags, BT_DEV_READY)) {
+	if (!IS_ENABLED(CONFIG_BT_CONN_LOOPBACK) &&
```
Comment (carlescufi) [inline subsys/bluetooth/host/conn.c]: why? Shouldn't we respect the device having to be ready before we create a connection? this might have additional side-effects otherwise if parts of `hci_core` are not ready?

**Review by carlescufi — COMMENTED** [review]
The addition of loopback is very welcome, and a nice way of detecting basic errors on the emulator, though ultimately and ideally we'd use nrf52_bsim instead, which tests the whole code path without instrumenting the code. Other than that I am OK with the changes except for some comments

File: include/bluetooth/gatt.h (hunk 2 above)
Comment (Vudentz) [reply include/bluetooth/gatt.h]: It can, it would just not be able to cancel it.

File: subsys/bluetooth/host/att.c [hunk 5]
```diff
@@ -71,6 +71,26 @@ NET_BUF_POOL_DEFINE(prep_pool, CONFIG_BT_ATT_PREPARE_COUNT, BT_ATT_MTU,
...
+		struct bt_gatt_write_params write;
+		struct bt_gatt_subscribe_params subscribe;
+#endif /* CONFIG_BT_GATT_CLIENT */
+	};
+};
+
+#define ATT_REQ_MAX_PARAMS \
+	(sizeof(struct bt_att_req) + sizeof(struct gatt_params))
+
+K_MEM_POOL_DEFINE(req_pool, sizeof(struct bt_att_req), ATT_REQ_MAX_PARAMS,
```
Comment (carlescufi) [inline subsys/bluetooth/host/att.c]: I am now wondering if this should be optional, since applications might be reusing the ATT request memory elsewhere and now we are dedicating it solely to this. It would complicate a bit the return values (0 or id, maybe we could make `id == 0` an invalid ID?)

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: It was because of the paramaters are copied otherwise the response will not be handled properly as it did not have the node entry set.

**Comment by Vudentz** [comment]
> The addition of loopback is very welcome, and a nice way of detecting basic errors on the emulator, though ultimately and ideally we'd use nrf52_bsim instead, which tests the whole code path without instrumenting the code. Other than that I am OK with the changes except for some comments

It is a protocol test facility with no external dependency nothing else.

File: subsys/bluetooth/host/att.c (hunk 5 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: From the feedback I got the maintaining of the requests on the application can get a little out of hand depending on the how many they are, and it is not the simple to detect if a request is in use, this actually makes testing a lot simpler since one can just make as many request as it wants without worrying if the parameters are not being overwritten or something. If I had to rewrite the test to have every single parameter stored in a static the tests would be somewhat different, or I would have to perform them twice one with parameters copy enable and another without which I don't thing is a complexity we'd like to have.

File: subsys/bluetooth/host/conn.c (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/host/conn.c]: We don't use the hci_core, in fact the test don't even have a driver.

File: subsys/bluetooth/host/gatt.c [hunk 6]
```diff
@@ -1383,29 +1383,38 @@ static void gatt_indicate_rsp(struct bt_conn *conn, u8_t err,
...
 		req->buf = buf;
 		req->func = func;
 		req->destroy = destroy;
 
-		err = bt_att_req_send(conn, req);
+		ret = bt_att_req_send(conn, req);
+		if (ret) {
+			bt_att_req_free(req);
+		} else {
+			ret = POINTER_TO_INT(req);
```
Comment (joerchan) [inline subsys/bluetooth/host/gatt.c]: This would be problematic for 64-bit system, (AFAIK we have some support for that now). Is there perhaps some form of ID that can be used to reference mem blocks? net_buf has something, but I couldn't find something for k_mem

File: include/bluetooth/gatt.h [hunk 7]
```diff
@@ -1244,6 +1226,7 @@ struct bt_gatt_subscribe_params {
 	u16_t value;
 	/** Subscription flags */
 	u8_t flags;
+	sys_snode_t *pnode;
```
Comment (joerchan) [inline include/bluetooth/gatt.h]: I don't understand this one, why do you need a pointer which always points to the object below?

**Review by joerchan — CHANGES_REQUESTED** [review]

File: include/bluetooth/gatt.h (hunk 7 above)
Comment (Vudentz) [reply include/bluetooth/gatt.h]: This is to refer to to the real params since this can be a copy used only for submitting the write.

File: subsys/bluetooth/host/gatt.c (hunk 6 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Not sure why would it be a problem on 64 bit, afaik int would be 64 bits as well.

File: subsys/bluetooth/host/conn.c [hunk 8]
```diff
@@ -1496,7 +1507,12 @@ struct bt_conn *bt_conn_add_le(const bt_addr_le_t *peer)
 	conn->sec_level = BT_SECURITY_LOW;
 	conn->required_sec_level = BT_SECURITY_LOW;
 #endif /* CONFIG_BT_SMP */
-	conn->type = BT_CONN_TYPE_LE;
+	if (IS_ENABLED(CONFIG_BT_CONN_LOOPBACK) &&
+	    !bt_addr_cmp(&peer->a, BT_ADDR_NONE)) {
+		conn->type = BT_CONN_TYPE_LOOPBACK;
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/conn.c]: I used the type exactly to avoid assuming some type, and doing address comparisons in the process.

File: subsys/bluetooth/host/gatt.c (hunk 6 above)
Comment (joerchan) [reply subsys/bluetooth/host/gatt.c]: No, int would be 32 bit. POINTER_TO_INT casts to uintptr_t which is defined as long unsigned int

File: subsys/bluetooth/host/conn.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply subsys/bluetooth/host/conn.c]: Right, but plenty of code does things differently depending on the role, for example SMP. But maybe loopback is not the best thing to test that with anyway.
Just a thought. This works well enough for GATT it seems.

File: include/bluetooth/gatt.h (hunk 7 above)
Comment (joerchan) [reply include/bluetooth/gatt.h]: Makes sense.
